---
title: "Multi Factor Analysis"
author: "MRC CSC Bioinformatics Core"
date: "08 August 2016"
output: html_document
Website: http://bioinformatics.csc.mrc.ac.uk/training.html
---

1. Use contrasts to specify the comparison you wish to make. Run comparisons between KOa vs FFa and KOb vs FFa. 

```{r,collapse=T,warning=FALSE,echo=FALSE,eval=TRUE,message=FALSE}

suppressPackageStartupMessages(library(DESeq2))

targetsE <- read.table("Exercise_ShortRNAseq_sample.info",sep="\t",header=TRUE)

AllCountsE<-read.csv(file="Exercise_ShortRNAseq_counts.csv",header=T,row.names=1)

cDataE<-data.frame(name=targetsE$sample,condition=targetsE$condition,batch=targetsE$batch)

ddsE<-DESeqDataSetFromMatrix(countData= AllCountsE,colData=cDataE,design=~batch+condition)
ddsE<-DESeq(ddsE)

```


```{r,eval=TRUE,echo=TRUE}    
 
    res_contrastE<-results(ddsE,contrast=c("condition","KOa","FFa")) 

    summary(res_contrastE)
    
    res_contrastE2<-results(ddsE,contrast=c("condition","KOb","FFa")) 

    summary(res_contrastE2)
      
```

2. Perform multi factor analysis by adding batch as factor to design.


a) Find the number of genes that are changed because of condition at FDR 0.05.

```{r, echo=T,warnings=F,collapse=T,eval=TRUE}

ddsE<-DESeqDataSetFromMatrix(countData=AllCountsE,colData=cDataE,design=~batch+condition)
dds_conditionE <- DESeq(ddsE)

resdds_conditionE<-results(dds_conditionE)

summary(resdds_conditionE,alpha=0.05)

```

b) Find the number of genes that are changed because of batch at FDR 0.05.

```{r, echo=T,warnings=F,collapse=T,eval=TRUE}


ddsE<-DESeqDataSetFromMatrix(countData=AllCountsE,colData=cDataE,design=~condition+batch)

dds_batchE <- DESeq(ddsE)

resdds_batchE<-results(dds_batchE)

summary(resdds_batchE,alpha=0.05)

```

3. Repeat above analysis using likelihood ratio test instead of the wald test,

a) Find the number of genes that are changed because of condition at FDR 0.05.




```{r, echo=T,warnings=F,collapse=T,eval=TRUE}

ddsE<-DESeqDataSetFromMatrix(countData=AllCountsE,colData=cDataE,design=~batch+condition)

ddsLRT_conditionE <- DESeq(ddsE, test="LRT", full=~batch+condition, reduced=~batch)

resddsLRT_conditionE<-results(ddsLRT_conditionE)

summary(resddsLRT_conditionE,alpha=0.05)

```

b) Find the number of genes that are changed because of batch at FDR 0.05.

```{r, echo=T,warnings=F,collapse=T,eval=TRUE}

# LRT analysis for the batch effect


ddsLRT_batchE <- DESeq(ddsE, test="LRT", full=~batch+condition, reduced=~condition)

resddsLRT_batchE<-results(ddsLRT_batchE)

summary(resddsLRT_batchE,alpha=0.05)

```

4. Introduce interaction in your design to to see if there is interaction between two factors involved in your design.

```{r,eval=TRUE,echo=TRUE}

ddsE$newgroup <- factor(paste0(ddsE$condition, ddsE$batch))
design(ddsE) <- ~ newgroup
ddsE <- DESeq(ddsE)
resultsNames(ddsE)
results(ddsE, contrast=c("newgroup", "FFaa", "KOaa"))

```








