---
title: "Visualization"
author: "MRC CSC Bioinformatics Core"
date: "08 August 2016"
output: html_document
Website: http://bioinformatics.csc.mrc.ac.uk/training.html
---

1. Transform the count data using rlog and vst method.

```{r,eval=TRUE,echo=TRUE}

rld<-rlog(dds)  

vsd <- varianceStabilizingTransformation(dds)

```

2. a) Perform the Principal Component Analysis using rlog method and find out % significance values of first two principal components .

```{r, echo=T,eval=FALSE}
 
suppressPackageStartupMessages(library(DESeq2))


plotPCA(rld, intgroup="condition")

```

 b) Repeat the PCA, this time using VST method and compare the plots with the ones obtained using rlog method.
 
```{r, echo=T,eval=FALSE}
 

plotPCA(vsd, intgroup="condition")

```
 
3. Perform clustering based on sample to sample distances and compare it with the clusters obtained using PCA.

```{r,eval=TRUE,echo=TRUE}   


rlogcount <- assay(rld)    
sampleDists <- as.matrix(dist(t(rlogcount)))

heatmap.2(as.matrix(sampleDists), key=F, trace="none",
    col=colorpanel(100, "black", "white"),
    ColSideColors=showcols[colData(dds)$Group], 
    RowSideColors=showcols[colData(dds)$Group],
    margin=c(10, 10), main="Sample Distance Matrix")
         
```

4. Draw a heatmap of count matrix based on top 40 differentially expressed genes using rlog and vst data.


```{r,eval=TRUE,echo=TRUE}

library("pheatmap")
select <- order(rowMeans(counts(dds,normalized=TRUE)),decreasing=TRUE)[1:20]


pheatmap(assay(rld)[select,])

pheatmap(assay(vsd)[select,])

```