---
title: "RNA-Seq exercise solutions"
author: "MRC CSC Bioinformatics Core"
date: "10/March/2016"
output: html_document
Website: http://bioinformatics.csc.mrc.ac.uk/training.html
---

In this exercise we will read in a count table containing counts from RNAseq experiment from erythroblast differentiation in mice This data was downloaded from GEO database (http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE49843) and aligned to mm9 by Rsubread. More details with regards to this experiment please refer to (http://www.ncbi.nlm.nih.gov/pubmed/24092935). We will perform differential expression analysis and find genes that were changed in knockdown samples versus control.

* Material

(1) Sample description: Exercise_ShortRNAseq_sample.info

(2) Count data: Exercise_ShortRNAseq_counts.csv

* set up the working directory

```{r,collapse=T,warning=FALSE}
# getwd()    # see your current directory
# setwd()    # set up your working directory
```

* First read in the sample and the count information.

* Also, please identify how many variables this experiment contains

```{r,collapse=T,warning=FALSE}

suppressPackageStartupMessages(library(DESeq2))

# load the sample description that we generated manually for this exercise

targets <- read.table("Exercise_ShortRNAseq_sample.info",sep="\t",header=TRUE)

# check the content of "targets" object that we just created

targets

```

In this experiment, we have two variables, namely, condition and batch. (ps: we introduced the 'mock' batch variable for this exercise. In the original publication, the batch effect was not discussed).

For the condition variable, we have 3 factors, namely FFa, KOa and KOb.

For the batch variable, we have 3 factors, namely a, b and c.

```{r,collapse=T,warning=FALSE}

# load the count information that we generated for this exercise by using Rsubread, featurecount function.

AllCounts<-read.csv(file="Exercise_ShortRNAseq_counts.csv",header=T,row.names=1)

# see the what is in the counts.csv

head(AllCounts)

# We provide entrez_id as identifier for this exercise

cData<-data.frame(name=targets$sample,condition=targets$condition,batch=targets$batch)

dds<-DESeqDataSetFromMatrix(countData= AllCounts,colData=cData,design=~batch+condition)

```

* Please perform the PCA analysis.

```{r, echo=T}
 
rld<-rlog(dds)

plotPCA(rld, intgroup="condition")

```

Now you have the count table as deseqdataset, you can start to perform the DE analysis.

* Find the number of genes that are changed in knockdown samples versus control, i.e. KOa vs FFa and KOb vs FFa, at FDR 0.05 irrespective of fold change.

* Find the number of genes that are changed in the above situation with fold change threshold, i.e. fold change ratio > 2.


```{r, echo=T,warnings=F,collapse=T}

dds<-DESeq(dds)

# shows how many genes were differentially expressed in KOa vs FFa, with FDR<0.05
res1<-results(dds, contrast=c("condition","KOa","FFa"))
summary(res1,alpha=0.05)

# shows how many genes were differentially expressed in KOa vs FFa, with FDR<0.05 and fold change ratio > 2
DE_res1<- res1[complete.cases(res1$padj),]
DE_res1<- DE_res1[DE_res1$padj<0.05 & abs(DE_res1$log2FoldChange)>1,]

# shows how many genes were differentially expressed in KOb vs FFa, with FDR<0.05
res2<-results(dds, contrast=c("condition","KOb","FFa"))

# shows how many genes were differentially expressed in KOa vs FFa, with FDR<0.05 and fold change ratio > 2
DE_res2<- res2[complete.cases(res2$padj),]
DE_res2<- DE_res2[DE_res2$padj<0.05 & abs(DE_res2$log2FoldChange)>1,]


```


* This time, use likelihood ratio test instead of the wald test.

* Find the number of genes that are changed because of condition at FDR 0.05 irrespective of fold change.

* Find the number of genes that are changed because of batch at FDR 0.05 irrespective of fold change.

```{r, echo=T,warnings=F,collapse=T}

ddsLRT<-DESeqDataSetFromMatrix(countData=AllCounts,colData=cData,design=~batch+condition)

# LRT analysis for the condition effect
ddsLRT_con <- DESeq(ddsLRT, test="LRT", full=~batch+condition, reduced=~batch) 

resddsLRT_con<-results(ddsLRT_con)

summary(resddsLRT_con,alpha=0.05)

resLRTorder<-resddsLRT_con[order(resddsLRT_con$padj),]

# LRT analysis for the batch effect

ddsLRT_batch <- DESeq(ddsLRT, test="LRT", full=~batch+condition, reduced=~condition)

resddsLRT_batch<-results(ddsLRT_batch)

summary(resddsLRT_batch,alpha=0.05)

```

Using the genes that are changed because of condition at FDR 0.05 irrespective of fold change as our differentially expressed genes, perform the Gene Ontology and Pathway Enrichment Analysis.

```{r, echo=T}

suppressPackageStartupMessages(library(KEGG.db))
suppressPackageStartupMessages(library(goseq))

# remove the NAs

resdat<- resLRTorder[complete.cases(resLRTorder$padj),]
 
degenes<-as.integer(resdat$padj<0.05)
names(degenes)<-rownames(resdat)
    
# remove duplicate gene names
degenes<-degenes[match(unique(names(degenes)),names(degenes))]
table(degenes)

# Fitting the probability weighting function (PWF)
# note, we use Entrez Gene ID as identifiler for this exercise
# we need to choose the correct "id" for the nullp function
# more details see
?nullp

pwf=nullp(degenes,'mm9','knownGene', plot.fit=FALSE)

# Calculate  the  over  and  under  expressed  GO categories among DE genes
go<-goseq(pwf,'mm9','knownGene', test.cats=c("GO:BP","GO:MF","KEGG"))

```

Change the Keggpath id to name in the goseq output

```{r,eval=TRUE,echo=TRUE}

# function that converts KEGG id to KEGG description
xx <- as.list(KEGGPATHID2NAME)
temp <- cbind(names(xx),unlist(xx))
    
addKeggTogoseq <- function(JX,temp){
  for(l in 1:nrow(JX)){
      if(JX[l,1] %in% temp[,1]){
          JX[l,"term"] <- temp[temp[,1] %in% JX[l,1],2]
          JX[l,"ontology"] <- "KEGG"
      }
  }
  return(JX)
}

restemp<-addKeggTogoseq(go,temp)   

head(restemp)

# save the goseq result
write.csv(restemp,file="Exercise_ShortRNAseq_GO_Kegg_Wallenius.csv", row.names=F)
```


